### ISSUE TITLE
Fast translate - is this the expected behaviour?

### ISSUE THREAD

**Issue by pschwllr** [OpenNMT/OpenNMT-py#995]
When using the `translate.py` script with the `-fast` option.

If the top beam predicts an end token, the end condition is matched and the rest of the beams for this particular sentence are not advanced anymore, e. g.:

Fast translate (top beam matches end condition, all beams finish):
BEST HYP:
[-23.4196] ['▁In', '▁diesem', '▁Zusammenhang', '▁stellt', '▁man', '▁fest', ',', '▁dass', '▁das', '▁Parlament', '▁in', '▁seine', 'm', '▁Interesse', '▁liegt', '.']
[-23.8467] ['▁In', '▁diesem', '▁Zusammenhang', '▁stellt', '▁man', '▁fest', ',', '▁dass', '▁das', '▁Parlament', '▁in', '▁seine', 'm', '▁Interesse', '▁liegt', '▁und', '▁dass']
[-24.3129] ['▁In', '▁diesem', '▁Zusammenhang', '▁stellt', '▁man', '▁fest', ',', '▁dass', '▁das', '▁Parlament', '▁in', '▁seine', 'm', '▁Interesse', '▁liegt', '▁und', '▁die']
[-24.3979] ['▁In', '▁diesem', '▁Zusammenhang', '▁stellt', '▁man', '▁fest', ',', '▁dass', '▁das', '▁Parlament', '▁in', '▁seine', 'm', '▁Interesse', '▁liegt', ',', '▁die']
[-24.7123] ['▁In', '▁diesem', '▁Zusammenhang', '▁stellt', '▁man', '▁fest', ',', '▁dass', '▁das', '▁Parlament', '▁in', '▁seine', 'm', '▁Interesse', '▁liegt', ',', '▁das']
PRED AVG SCORE: -1.4637, PRED PPL: 4.3220

Original “slow” translate (beams are continued until end token predicted): 
BEST HYP:
[-23.4196] ['▁In', '▁diesem', '▁Zusammenhang', '▁stellt', '▁man', '▁fest', ',', '▁dass', '▁das', '▁Parlament', '▁in', '▁seine', 'm', '▁Interesse', '▁liegt', '.']
[-35.4067] ['▁In', '▁diesem', '▁Zusammenhang', '▁stellt', '▁man', '▁fest', ',', '▁dass', '▁das', '▁Parlament', '▁in', '▁seine', 'm', '▁Interesse', '▁liegt', '▁und', '▁dass', '▁es', '▁sich', '▁hier', '▁um', '▁eine', '▁Entscheidung', '▁handelt', '.']
[-46.6008] ['▁In', '▁diesem', '▁Zusammenhang', '▁stellt', '▁man', '▁fest', ',', '▁dass', '▁das', '▁Parlament', '▁in', '▁seine', 'm', '▁Interesse', '▁liegt', '▁und', '▁dass', '▁es', '▁sich', '▁hierbei', '▁um', '▁eine', '▁Entscheidung', '▁handelt', ',', '▁die', '▁von', '▁der', '▁Bedeutung', '▁des', '▁Parlament', 's', '▁ab', 'weich', 't', '.']
[-46.6126] ['▁In', '▁diesem', '▁Zusammenhang', '▁stellt', '▁man', '▁fest', ',', '▁dass', '▁das', '▁Parlament', '▁in', '▁seine', 'm', '▁Interesse', '▁liegt', '▁und', '▁dass', '▁es', '▁sich', '▁hierbei', '▁um', '▁eine', '▁Entscheidung', '▁handelt', ',', '▁die', '▁von', '▁der', '▁Bedeutung', '▁des', '▁Parlament', 's', '▁abhäng', 't', '.']
[-57.1445] ['▁In', '▁diesem', '▁Zusammenhang', '▁stellt', '▁man', '▁fest', ',', '▁dass', '▁das', '▁Parlament', '▁in', '▁seine', 'm', '▁Interesse', '▁liegt', '▁und', '▁dass', '▁es', '▁sich', '▁hierbei', '▁um', '▁eine', '▁Entscheidung', '▁handelt', ',', '▁die', '▁von', '▁der', '▁Bedeutung', '▁des', '▁Parlament', 's', '▁ab', 'weich', 't', '▁und', '▁in', '▁dessen', '▁Interesse', '▁liegt', '.']
PRED AVG SCORE: -1.4637, PRED PPL: 4.3220

Therefore, only the top-1 prediction will for sure correspond to the “slow” translate results. Top-n (n>1) predictions will be cut at the point the end condition was matched. 

I would suggest to add an `assert n_best == 1` to the `_fast_translate_batch` function.

**Comment by guillaumekln** [comment]
Indeed, a batch is currently dropped as soon as the top beam reaches the end. I will push a fix to continue the search until `n_best` finished hypotheses can be returned.

**Comment by pschwllr** [comment]
You're right, to have it `n-best` would be even better. Many thanks!
